nexus/reconfigurator/simulation/src/sim.rs

```diff
@@ -0,0 +1,146 @@
+/// cloning. In the future, it would be interesting to store data at the top
+/// level, and only store references to it within each `SimState`. (The
+/// references could be UUIDs, Rust `&` references, or content-addressed
+/// hashes.) This does add some complexity, though, and there isn't a clear
+/// benefit at the moment. For more, see the comment within the definition of
+/// `system.rs`'s `SimSystem`.
+#[derive(Clone, Debug)]
+pub struct Simulator {
+    log: slog::Logger,
+    // In the future, it would be interesting to store a chain of every set of
+    // heads over time, similar to `jj op log`. That would let us implement undo
+    // and restore operations.
+    heads: IndexSet<ReconfiguratorSimUuid>,
+    states: HashMap<ReconfiguratorSimUuid, Arc<SimState>>,
+    // This state corresponds to `ROOT_ID`.
+    //
+    // Storing it in the Arc is extremely important! `SimStateBuilder` stores a
+    // pointer to the root state to ensure that there's no attempt to compare
+    // "unrelated histories". The Simulator struct itself can both be cloned and
+    // moved in memory, but because `root_state` is never changed, it always
+    // points to the same memory address.
+    root_state: Arc<SimState>,
+    // Top-level (unversioned) RNG.
+    sim_uuid_rng: TypedUuidRng<ReconfiguratorSimKind>,
+}
+
+impl Simulator {
+    /// The root ID of the store.
+    ///
+    /// This is always defined to be the nil UUID, and if queried will always
+    /// have a state associated with it.
+    pub const ROOT_ID: ReconfiguratorSimUuid = ReconfiguratorSimUuid::nil();
+
+    /// Create a new simulator with the given initial seed.
+    pub fn new(log: &slog::Logger, seed: Option<String>) -> Self {
+        let seed = match seed {
+            Some(seed) => seed,
+            None => seed_from_entropy(),
+        };
+        Self::new_inner(log, seed)
+    }
+
+    fn new_inner(log: &slog::Logger, seed: String) -> Self {
+        let log = log.new(slog::o!("component" => "SimStore"));
+        let sim_uuid_rng =
+            TypedUuidRng::from_seed(&seed, "ReconfiguratorSimUuid");
+        let root_state = SimState::new_root(seed);
+        Self {
+            log,
+            heads: IndexSet::new(),
+            states: HashMap::new(),
+            root_state,
+            sim_uuid_rng,
+        }
+    }
+
+    /// Get the initial RNG seed.
+    ///
+    /// Versioned configurations start with this seed, though they may choose
+    /// to change it as they go along.
+    pub fn initial_seed(&self) -> &str {
+        &self.root_state.rng().seed()
+    }
+
+    /// Get the current heads of the store.
+    #[inline]
+    pub fn heads(&self) -> &IndexSet<ReconfiguratorSimUuid> {
+        &self.heads
+    }
+
+    /// Get the state for the given UUID.
+    pub fn get_state(&self, id: ReconfiguratorSimUuid) -> Option<&SimState> {
+        if id == Self::ROOT_ID {
+            return Some(&self.root_state);
```

Comment:
This seems fine, but out of curiosity - why not insert the root state into `self.states`?

Reply:
Seemed cleaner to me that way -- it could live in `self.states` instead of being its own member.